### ISSUE TITLE
program runs ok, but accuracy does not improve over epoch

### ISSUE THREAD

**Issue by hjk41** [hjk41/MxNet.cpp#12]
I am using the new optimizer interface to write a very simple program which basically uses two FC layer to predict Ads CTR. The program runs ok, but the accuracy does not improve at all across epoch. I suspect I am doing something stupid, but I can't figure it out.

Could you help take a look? 

https://github.com/hjk41/MxNet.cpp/blob/ads/windows/vs/MxnetTestApp/train_ads.cpp

**Comment by zhangchen-qinyinghua** [comment]
Ok, I will check the code.
Hope it is not some hidden bug...

**Comment by hjk41** [comment]
@zhangchen-qinyinghua  I think I find the bug. Executor::outputs never get assigned... So the prediction result always stays the same. I think we should get the outputs in Executor::Forward()

**Comment by hjk41** [comment]
One thing I notice is that I didn't bind the weights and biases explicitly. After looking at SimpleBind, I think that caused some problem, since the weights and biases will be reinitialized after each SimpleBind. However, after fixing the problem, the outputs still remains zero all the time. I will need more time to figure it out

**Comment by zhangchen-qinyinghua** [comment]
I think the problem is that before using infer the args_map you should tell MxNet.cpp some information about the shape.
Some thing like:
'''
args_map["data"] = NDArray(context,shape);
args_map["label"] = NDArray(context,shape);
'''
Before https://github.com/hjk41/MxNet.cpp/blob/ads/windows/vs/MxnetTestApp/train_ads.cpp#L49.

And SimpleBind will only initialize the weights and biases if they are not initialized before.
Codes here:
https://github.com/hjk41/MxNet.cpp/blob/master/include/symbol.hpp#L162-L166

**Comment by hjk41** [comment]
I think this will create a new empty NDArray and use it as the parameter,
no?

On Thu, Jan 21, 2016 at 10:26 AM, Zhang Chen <notifications@github.com>
wrote:

> I think the problem is that before using infer the args_map you should
> tell MxNet.cpp some information about the shape.
> Some thing like:
> '''
> args_map["data"] = NDArray(context,shape);
> args_map["label"] = NDArray(context,shape);
> '''
> Before
> https://github.com/hjk41/MxNet.cpp/blob/ads/windows/vs/MxnetTestApp/train_ads.cpp#L49
> .
>
> And SimpleBind will only initialize the weights and biases if they are not
> initialized before.
> Codes here:
> https://github.com/hjk41/MxNet.cpp/blob/master/include/symbol.hpp#L162-L166
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hjk41/MxNet.cpp/issues/12#issuecomment-173431746>.
>



-- 
HONG Chuntao
System Research Group
Microsoft Research Asia

**Comment by zhangchen-qinyinghua** [comment]
InferArgsMap creates new empty NDArray from known parameters, only with some known parameters can it infer shapes of new NDArrays.
So the shapes of args_map["data"] and args_map["label"] must be given.

**Comment by zhangchen-qinyinghua** [comment]
Here, I think it is also OK to just remove this line:
https://github.com/hjk41/MxNet.cpp/blob/ads/windows/vs/MxnetTestApp/train_ads.cpp#L49

I will write a mlp example with new operator wrappers, and fix bugs if I find some.

**Comment by hjk41** [comment]
It works now. I added some lines in Executor::Forward, do you think it is necessary?
```c++
  void Forward(bool is_train) {
    MXExecutorForward(handle_, is_train ? 1 : 0);
    mx_uint out_size;
    NDArrayHandle *out_array;
    CHECK_EQ(MXExecutorOutputs(handle_, &out_size, &out_array), 0);
    for (mx_uint i = 0; i < out_size; ++i) {
      outputs[i] = NDArray(out_array[i]);
    }
  }
```

**Comment by zhangchen-qinyinghua** [comment]
@hjk41 
I don't think this is necessary. Because I already bind the outputs array when the executor is constructed:
https://github.com/hjk41/MxNet.cpp/blob/master/include/executor.hpp#L50-L54

I do this just as the python interface does. But I never test this in detail.

**Comment by hjk41** [comment]
The outputs (NDArrayHandle) I got from the executors varies each time. So I
think it might be necessary to update the NDArrayHandles. Let's dig deeper
into the code.

On Thu, Jan 21, 2016 at 5:28 PM, Zhang Chen <notifications@github.com>
wrote:

> @hjk41 <https://github.com/hjk41>
> I don't think this is necessary. Because I already bind the outputs array
> when the executor is constructed:
> https://github.com/hjk41/MxNet.cpp/blob/master/include/executor.hpp#L50-L54
>
> I do this just as the python interface does. But I never test this in
> detail.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hjk41/MxNet.cpp/issues/12#issuecomment-173513668>.
>



-- 
HONG Chuntao
System Research Group
Microsoft Research Asia

**Comment by zhangchen-qinyinghua** [comment]
I see.
I will try to test it.
